File: computer_thinking/rust_problem_solving/src/maths/armstrong.rs

// Time: theta(log N) where N is the number, log N is number of digits
// Space: theta(1)
pub fn is_armstrong(mut n: u32) {
    let original = n;
    let mut num = n;
    let mut count = 0u32;
    let mut sum = 0u32;
    const BASE_10: u32 = 10;

    while n > 0 {
        n = n / BASE_10;
        count +=1;
    }

    while num > 0{
        let digit = num % BASE_10;
        sum = sum + digit.checked_pow(count).expect("Shouldn't overflow when calculating sum");
        num = num / BASE_10;
    }

    match sum == original {
        true => println!("Yes, it's an Armstrong number"),
        false => println!("No, it's not an Armstrong number")
    }
}

File: python_dsa/maths/armstrong.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
)


// Time: theta(logN) where N is integer
// Aux space: theta(1)
func armstrongNumber(n int) {
	num := n
	original := n
	count := 0
	// find number of digits in number is base-10
	
	for n > 0 {
		n = n / 10
		count += 1
	}

	// calculate sum of pow(digit, no.of digits) for all digits
	sum := 0
	for num > 0{
		digit := num % 10
		sum += int(math.Pow(float64(digit), float64(count)))
		num = num /10
	}

	// check if sum == n
	if sum == original {
		fmt.Printf("Yes, it's an Armstrong Number\n")
	} else {
		fmt.Printf("No, it is not an Armstrong Number\n")
	}

}

func main() {
	args := os.Args[1:]
	if num, err := strconv.Atoi(args[0]); err == nil {
		armstrongNumber(num)
	}
}
